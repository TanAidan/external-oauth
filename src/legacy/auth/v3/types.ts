import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
