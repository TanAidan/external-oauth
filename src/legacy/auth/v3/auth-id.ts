import { Response, NextFunction } from 'express'
import { TAuthIdRequest } from './types'
import { v1 as uuidv1 } from 'uuid'

export const connectAuthId = (req: TAuthIdRequest, res: Response, next: NextFunction) => {
  const currentAuthId = req.query.authId as string
  const authId = currentAuthId || uuidv1()

  req.authId = authId
  req.session.authId = authId

  next()
}

export const callbackAuthId = (req: TAuthIdRequest, _res: Response, next: NextFunction) => {
  const { authId } = req.session
  req.authId = authId

  next()
}
