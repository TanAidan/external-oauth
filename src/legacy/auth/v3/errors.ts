import { UserError } from '../../errors'
import { EAuthType } from './types'
import { Request } from 'express'
import { GrantType } from './strategies/oauth2'

export class NoAuthInProgress extends UserError {
  constructor() {
    super(
      `The callback was invoked whilst no authentication was in progress. ` +
        `Either too much time has elapsed between starting authentication and the callback being invoked, ` +
        `or the callback was invoked without first calling the integration's authentication endpoint. Please try again`,
      422,
      'NO_AUTH_IN_PROGRESS'
    )
  }
}
export class InvalidAuthType extends UserError {
  constructor(authType: string) {
    const allowedValues = Object.values(EAuthType)
      .map(authType => `'${authType}'`)
      .join(', ')
    super(
      `Invalid authType '${authType}'. Please check the integration's auth.config.json file. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_AUTH_TYPE'
    )
  }
}

export class InvalidGrantType extends UserError {
  constructor(grantType: string) {
    const allowedValues = Object.values(GrantType)
      .map(grantType => `'${grantType}'`)
      .join(', ')
    super(
      `Invalid grantType '${grantType}'. Please check the API's configuration. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_GRANT_TYPE'
    )
  }
}

export class OAuthOnlyEndpoint extends UserError {
  constructor(authType: string) {
    super(
      "This endpoint is only for 'OAUTH1' and 'OAUTH2' auth types " +
        `but the integration is configured with authType '${authType}'`,
      422,
      'OAUTH_ONLY_ENDPOINT'
    )
  }
}

export class AuthenticationFailed extends UserError {
  readonly error: string
  readonly errorDescription?: string

  static fromOAuthRequest(req: Request, info: any) {
    if (req.query.error) {
      return new AuthenticationFailed(req.query)
    }

    return new AuthenticationFailed(info)
  }

  constructor(details: any) {
    super(`Authentication failed:\n${JSON.stringify(details, undefined, 2)}`, 403, 'AUTHENTICATION_FAILED')

    if (details.error) {
      this.error = details.error
      this.errorDescription = details.error_description
    } else {
      this.errorDescription = this.error = JSON.stringify(details)
    }
  }
}

export class AccessTokenExpired extends UserError {
  constructor() {
    super(
      `The access token associated with your Auth Id has expired and token refreshing is not supported for \
this API. Please reconnect to obtain a new access token.

Consult the API provider's documentation to determine whether refresh tokens can be enabled for this API`,
      403,
      'ACCESS_TOKEN_EXPIRED'
    )
  }
}

export class InconsistentSetupId extends UserError {
  constructor() {
    super(
      'The provided setupId does not match the setupId used when connecting. ' +
        'Please reconnect using this setupId, or retry this call with the ' +
        'setupId corresponding to the authId',
      422,
      'INCONSISTENT_SETUP_ID'
    )
  }
}

export class InvalidAuthId extends UserError {
  constructor(buid: string, authId: string) {
    super(
      `No auth details were found using authId '${authId}' for API '${buid}'

Please try again with a valid authId or connect with this authId`,
      404,
      'INVALID_AUTH_ID'
    )
  }
}

export class MissingAuthId extends UserError {
  constructor(buid: string) {
    super(
      `You must supply an authId to use the '${buid}' API

Please try again with a valid authId.

See the following link for information on how to obtain an authId: https://docs.bearer.sh/faq/connect-button`,
      401,
      'MISSING_AUTH_ID'
    )
  }
}

export class CredentialsNotConfigured extends UserError {
  constructor(buid: string) {
    super(
      `Missing credentials for the '${buid}' API

Please configure the credentials in the dashboard and try again.

See the following link for information: https://docs.bearer.sh/dashboard/apis`,
      422,
      'CREDENTIALS_NOT_CONFIGURED'
    )
  }
}
