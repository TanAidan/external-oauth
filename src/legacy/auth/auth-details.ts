import { Response, NextFunction } from 'express'

import { AuthDetailsRequest } from './configure-request'
import { AuthDetails, EAuthType, OAuth1AuthDetails, OAuth2AuthDetails } from './v3/types'
import { OAuthOnlyEndpoint } from './v3/errors'
import { isOAuthType } from './v3/strategy'
import { asyncMiddleware } from '../errorHandler'
import { NO_VALUE } from './v3/strategies/oauth2/common'

enum TokenType {
  OAuth1 = 'oauth',
  OAuth2AccessToken = 'bearer',
  OAuth2RefreshToken = 'refresh', // Not defined in RFC7662
  OpenIDConnectToken = 'id' // Not defined in RFC7662
}

// Follows RFC7662 unless stated otherwise
const formatToken = ({
  clientId,
  expiresIn,
  scopes,
  tokenType,
  updatedAt,
  value
}: {
  clientId: string
  expiresIn?: number
  scopes?: string[]
  tokenType: TokenType
  updatedAt: number
  value?: string
}) => {
  if (!value || value === NO_VALUE) {
    return
  }

  const issuedAt = Math.trunc(updatedAt / 1000)
  const expiresAt = expiresIn ? issuedAt + expiresIn : undefined
  const active = !expiresAt || expiresAt * 1000 > Date.now()

  return {
    active,
    value, // Not defined in RFC7662
    client_id: clientId,
    exp: expiresAt,
    iat: issuedAt,
    scope: scopes ? scopes.join(' ') : undefined,
    token_type: tokenType
  }
}

const formatAuthDetails = (authType: EAuthType, authDetails: OAuth1AuthDetails | OAuth2AuthDetails) => {
  if (authType === EAuthType.OAuth1) {
    const {
      accessToken,
      callbackParams,
      consumerKey,
      consumerSecret,
      expiresIn,
      signatureMethod,
      tokenResponse,
      tokenSecret,
      updatedAt
    } = authDetails as OAuth1AuthDetails

    return {
      callbackParams,
      consumerKey,
      consumerSecret,
      signatureMethod,
      tokenResponse,
      tokenSecret,
      accessToken: formatToken({
        expiresIn,
        updatedAt,
        clientId: consumerKey,
        tokenType: TokenType.OAuth1,
        value: accessToken
      })
    }
  }

  const {
    accessToken,
    callbackParams,
    clientId,
    clientSecret,
    expiresIn,
    idToken,
    idTokenJwt,
    refreshToken,
    tokenResponse,
    updatedAt
  } = authDetails as OAuth2AuthDetails

  // OAuth2
  return {
    callbackParams,
    clientId,
    clientSecret,
    idTokenJwt,
    tokenResponse,
    accessToken: formatToken({
      clientId,
      updatedAt,
      expiresIn,
      tokenType: TokenType.OAuth2AccessToken,
      value: accessToken
    }),
    idToken: formatToken({ clientId, updatedAt, tokenType: TokenType.OpenIDConnectToken, value: idToken }),
    refreshToken: formatToken({ clientId, updatedAt, tokenType: TokenType.OAuth2RefreshToken, value: refreshToken })
  }
}

export const authDetailsResponse = asyncMiddleware(
  async (req: FetchAuthDetailsRequest, res: Response, _next: NextFunction) => {
    const { authType } = await req.integration.config()

    if (!isOAuthType(authType)) {
      throw new OAuthOnlyEndpoint(authType)
    }

    res.json(formatAuthDetails(authType, req.auth as OAuth1AuthDetails | OAuth2AuthDetails))
  }
)

export interface FetchAuthDetailsRequest extends AuthDetailsRequest {
  auth: AuthDetails
}
