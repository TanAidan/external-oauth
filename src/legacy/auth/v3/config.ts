import { asyncMiddleware } from '../../errorHandler'
import { IAuthConfig, EAuthType, TConnectConfigRequest, TCallbackConfigRequest } from './types'
import { NextFunction, Response } from 'express'
import {
  getConfig
  // getSetupDetails
} from '../clients/integrations'
import { InvalidAuthType } from './errors'
import { expandAuthConfig } from '../../api-config/auth-config'

const getAuthConfig = async (req: TConnectConfigRequest) => {
  const { connectParams, buid, configuration } = req

  return {
    setupDetails: configuration,
    integrationConfig: expandAuthConfig({
      connectParams,
      authConfig: (await getConfig({
        buid: buid!
      })) as any
    })
  } as IAuthConfig
}

const copyConfig = (src: IAuthConfig, dest: IAuthConfig) => {
  dest.integrationConfig = src.integrationConfig
  dest.setupDetails = src.setupDetails
}

export const connectConfig = asyncMiddleware(async (req: TConnectConfigRequest, _res: Response, next: NextFunction) => {
  const authConfig = await getAuthConfig(req)
  const { authType } = authConfig.integrationConfig

  if (!Object.values(EAuthType).includes(authType)) {
    throw new InvalidAuthType(authType)
  }

  copyConfig(authConfig, req)
  req.session.authConfig = authConfig

  next()
})

export const callbackConfig = (req: TCallbackConfigRequest, _res: Response, next: NextFunction) => {
  copyConfig(req.session.authConfig, req)

  next()
}
