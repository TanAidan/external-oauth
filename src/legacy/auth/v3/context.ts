import { TConnectContextRequest, TCallbackContextRequest } from './types'
import { NextFunction, Response } from 'express'
import { UserError } from '../../errors'
import { NoAuthInProgress } from './errors'
import { asyncMiddleware } from '../../errorHandler'

const getConnectParams = (params?: any) => {
  if (params === undefined) {
    return {}
  }

  if (typeof params !== 'object' || Array.isArray(params)) {
    throw new InvalidConnectParams()
  }

  for (const [name, value] of Object.entries(params)) {
    if (typeof value !== 'string' || !/^[\w\s.-]*$/.test(value)) {
      throw new InvalidConnectParam(name)
    }
  }

  return params
}

export const connectContext = (req: TConnectContextRequest, res: Response, next: NextFunction) => {
  const { buid, setupId } = req

  const connectParams = getConnectParams(req.query.params)

  req.session.context = {
    connectParams,
    setupId,
    buid
  }

  req.connectParams = connectParams

  next()
}

export const callbackContext = asyncMiddleware(
  async (req: TCallbackContextRequest, _res: Response, next: NextFunction) => {
    if (!req.session.context) {
      throw new NoAuthInProgress()
    }

    const { buid, connectParams } = req.session.context

    req.isCallback = true
    req.connectParams = connectParams
    req.buid = buid
    req.setupId = req.session.authConfig.setupDetails.setup_id

    next()
  }
)

class InvalidConnectParams extends UserError {
  constructor() {
    super(
      // tslint:disable:max-line-length
      `Incorrect format for connect parameters'

Connect parameters must be sent as query parameters of the form \`params[name]=value\` eg. \`params[subdomain]=my-app\`'`,
      // tslint:enable:max-line-length
      400,
      'INVALID_CONNECT_PARAMS'
    )
  }
}

class InvalidConnectParam extends UserError {
  constructor(name: string) {
    super(
      `Incorrect format for connect parameter '${name}'

Connect parameters may contain alphanumeric and space characters, or any of the following symbols '_-.'

Refer to this link for further information: https://docs.bearer.sh/faq/connect-button`,
      400,
      'INVALID_CONNECT_PARAM'
    )
  }
}
