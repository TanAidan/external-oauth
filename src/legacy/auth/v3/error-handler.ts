import { TErrorHandlerRequest } from './types'
import { Response, NextFunction } from 'express'
import { AuthenticationFailed, NoAuthInProgress } from './errors'
import { isEmpty } from '../../functions/utils'

const respondWithOAuthError = (
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/error', { error: `${code}: ${message}` })
}

const respondWithCallbackError = (req: TErrorHandlerRequest, res: Response, err: AuthenticationFailed) => {
  res.header('Content-Type', 'text/html')
  res.status(err.statusCode)
  res.render('auth/callback', {
    authId: req.authId,
    error: err.error,
    error_description: err.errorDescription,
    integrationUuid: req.buid
  })
}

const respondWithCallbackPlaceholder = (
  req: TErrorHandlerRequest,
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/placeholder', {
    error: `${code}: ${message}`,
    callbackUrl: req.originalUrl,
    method: req.method
  })
}

export const errorHandler = (err: any, req: TErrorHandlerRequest, res: Response, next: NextFunction) => {
  if (res.headersSent) {
    return next(err)
  }

  if (err.statusCode) {
    if (err instanceof AuthenticationFailed) {
      return respondWithCallbackError(req, res, err)
    }

    if (err instanceof NoAuthInProgress && isEmpty(req.query)) {
      return respondWithCallbackPlaceholder(req, res, err)
    }

    return respondWithOAuthError(res, err)
  }
  respondWithOAuthError(res, {
    statusCode: 500,
    code: 'INTERNAL_ERROR',
    message: 'Encountered an unexpected error. Please contact support'
  })
}
