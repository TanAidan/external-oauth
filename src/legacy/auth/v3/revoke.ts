import { asyncMiddleware } from '../../errorHandler'
import { TRevokeRequest } from './types'
// import { revokeAuthV3 } from '../../clients/integrations'
import { Response } from 'express'
import { MissingParameter } from '../../errors'

export const revoke = asyncMiddleware(async (req: TRevokeRequest, res: Response) => {
  // const { authId } = req.params
  // const { buid } = req
  const { clientId } = req.query

  if (!clientId) {
    throw new MissingParameter('clientId')
  }

  // await revokeAuthV3({
  //   clientId,
  //   buid,
  //   authId,
  //   servicesTableName: req.stageVariables.servicesTableId
  // })

  res.status(200).end()
})
