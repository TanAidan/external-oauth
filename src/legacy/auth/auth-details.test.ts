import timekeeper from 'timekeeper'

import { MiddlewareTestHarness } from '../../../tests/utils'
import { authDetailsResponse, FetchAuthDetailsRequest } from './auth-details'
import { EAuthType } from './v3/types'
import { OAuthOnlyEndpoint } from './v3/errors'

describe('authDetailsResponse', () => {
  const commonDetails = {
    callbackParams: { callback: 'param' },
    expiresIn: 1234,
    tokenResponse: { token: 'response' },
    updatedAt: 1000
  }

  const oauth1AuthDetails = {
    ...commonDetails,
    accessToken: 'test-access-token',
    consumerKey: 'test-consumer-key',
    consumerSecret: 'test-consumer-secret',
    signatureMethod: 'PLAINTEXT',
    tokenSecret: 'test-token-secret'
  }

  const oauth2AuthDetails = {
    ...commonDetails,
    accessToken: 'test-access-token',
    clientId: 'test-client-id',
    clientSecret: 'test-client-secret',
    idToken: 'test-id-token',
    idTokenJwt: { jwt: 'token' },
    refreshToken: 'test-refresh-token'
  }

  const setup = ({ authType = EAuthType.OAuth2, withAuth = true } = {}) =>
    new MiddlewareTestHarness<FetchAuthDetailsRequest>({
      configureRequest: req => {
        req.integration = { config: () => Promise.resolve({ authType }) } as any
        req.authId = 'test-auth-id'
        req.buid = 'test-alias'

        if (withAuth) {
          req.auth = (authType === EAuthType.OAuth1 ? oauth1AuthDetails : oauth2AuthDetails) as any
        }
      },
      testMiddleware: authDetailsResponse
    })

  it('returns formatted data for OAuth1 APIs', async () => {
    const test = setup({ authType: EAuthType.OAuth1 })

    const res = await test.get().expect(200)

    expect(res.body).toMatchSnapshot()
  })

  it('returns formatted data for OAuth1 APIs', async () => {
    const test = setup({ authType: EAuthType.OAuth2 })

    const res = await test.get().expect(200)

    expect(res.body).toMatchSnapshot()
  })

  it('returns true for the access token active value when the "expires at" value is in the future', async () => {
    await timekeeper.withFreeze(1234, async () => {
      const test = setup()

      const res = await test.get().expect(200)

      expect(res.body.accessToken.active).toBeTruthy()
    })
  })

  it('raises an OAuthOnlyEndpoint error if the API auth type is not oauth1/2', async () => {
    const test = setup({ authType: EAuthType.NoAuth })

    await test.get().expect(422)

    expect(test.err).toBeInstanceOf(OAuthOnlyEndpoint)
  })
})
