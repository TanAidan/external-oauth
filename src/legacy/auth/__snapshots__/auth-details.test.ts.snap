// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authDetailsResponse returns formatted data for OAuth1 APIs 1`] = `
Object {
  "accessToken": Object {
    "active": false,
    "client_id": "test-consumer-key",
    "exp": 1235,
    "iat": 1,
    "token_type": "oauth",
    "value": "test-access-token",
  },
  "callbackParams": Object {
    "callback": "param",
  },
  "consumerKey": "test-consumer-key",
  "consumerSecret": "test-consumer-secret",
  "signatureMethod": "PLAINTEXT",
  "tokenResponse": Object {
    "token": "response",
  },
  "tokenSecret": "test-token-secret",
}
`;

exports[`authDetailsResponse returns formatted data for OAuth1 APIs 2`] = `
Object {
  "accessToken": Object {
    "active": false,
    "client_id": "test-client-id",
    "exp": 1235,
    "iat": 1,
    "token_type": "bearer",
    "value": "test-access-token",
  },
  "callbackParams": Object {
    "callback": "param",
  },
  "clientId": "test-client-id",
  "clientSecret": "test-client-secret",
  "idToken": Object {
    "active": true,
    "client_id": "test-client-id",
    "iat": 1,
    "token_type": "id",
    "value": "test-id-token",
  },
  "idTokenJwt": Object {
    "jwt": "token",
  },
  "refreshToken": Object {
    "active": true,
    "client_id": "test-client-id",
    "iat": 1,
    "token_type": "refresh",
    "value": "test-refresh-token",
  },
  "tokenResponse": Object {
    "token": "response",
  },
}
`;
